elastic/beats
# Remote global hostfs variables

## What does this PR do?

This is an attempt at cleaning up and removing the global `system.hostfs` flag that's been a source of pain for a while. This is a breaking change, but it's needed as we move the codebase forward, as the current management of 'hostfs` as a config concept is woefully ill-fitting for Agent's remote management.

There's a few things going on here:
- Remove the (rather confusingly named) `system.hostfs` command-line flag.
- Remove `hostfs` from the global `paths` library
- Remove the hacks that we used to get the `paths` hostfs management to play nice with agent
- Fix any docs/config files

Note that there's still code here that's fundamentally buggy, or at least weird, as we need to set a series of global environment variables for vendor libraries. This, by design, doesn't really play nice with how agent manages config. Fixing this will require wider refactoring of the system module, which is currently underway. 

I'm opening this as a draft PR, as I'm still in the process of testing it, and I haven't thoroughly gone over the docs.

## Why is it important?

We've run into a few bugs that have resulted from Agent's central management trying to play nice with the global, beat wide `system.hostfs` setting. In many ways it's awkward and unnecessary, and treating it as a module-level variable is more in line with an "agent-first" architecture for metricbeat as a whole.

## Checklist



## How to test this PR locally

This is honestly a nightmare to test. The best way is using Docker with `elastic-agent`:

- after you've pulled down the changes, run `PLATFORMS=linux/amd64 PACKAGES=docker mage dev:package` from within the `x-pack/elastic-agent` directory.
- If you're using `elastic-package`, change the dockerfile in `~/.elastic-package/profiles/default/stack/snapshot.yml` to update the `elastic-agent` container (note the `image` name and the bindmount):
```
  elastic-agent:
    image: docker.elastic.co/beats/elastic-agent-complete:8.1.0
    depends_on:
      fleet-server:
        condition: service_healthy
    healthcheck:
      test: "elastic-agent status"
      retries: 90
      interval: 1s
    hostname: docker-fleet-agent
    environment:
    - "FLEET_ENROLL=1"
    - "FLEET_INSECURE=1"
    - "FLEET_URL=http://fleet-server:8220"
    - "STATE_PATH=/usr/share/elastic-agent"
    volumes:
    - type: bind
      source: ../../../tmp/service_logs/
      target: /tmp/service_logs/
    - type: bind
      source: /proc
      target: /hostfs/proc
```

After that, spin up `elastic-package stack` 

- Set an alternate hostfs in the integration system config as `/hostfs`
- Make sure we're still ingesting data
